python/src/wslink/websocket.py

```diff
@@ -118,331 +111,10 @@ def updateSecret(self, newSecret):
-                    for i, v in enumerate(o):
-                        o[i] = findAttachments(v)
-                elif isinstance(o, dict):
-                    for k in o:
-                        o[k] = findAttachments(o[k])
-                return o
-            args = findAttachments(args)
-            kwargs = findAttachments(kwargs)
-
-            args.insert(0, obj)
-            results = maybeDeferred(func, *args, **kwargs)
-
-            results.addCallback(createSuccessCallback(self, rpcid, methodName))
-            results.addErrback(createErrorCallback(self, rpcid, methodName))
-        except Exception as e:
-            self.sendWrappedError(rpcid, EXCEPTION_ERROR, "Exception raised",
-                { "method": methodName, "exception": repr(e), "trace": traceback.format_exc() })
-            return
-
-
-    def sendWrappedMessage(self, rpcid, content, method=''):
-        wrapper = {
-            "wslink": "1.0",
-            "id": rpcid,
-            "result": content,
-        }
-        try:
-            encMsg = json.dumps(wrapper, ensure_ascii = False).encode('utf8')
-        except TypeError as e:
-            # the content which is not serializable might be arbitrarily large, don't include.
-            # repr(content) would do that...
-            self.sendWrappedError(rpcid, RESULT_SERIALIZE_ERROR, "Method result cannot be serialized", method)
-            return
-
-        # Check if any attachments in the map go with this message
-        attachments = publishManager.getAttachmentMap()
-        if attachments:
-            for key in attachments:
-                # string match the encoded attachment key
-                if key.encode('utf8') in encMsg:
-                    # send header
-                    header = {
-                        "wslink": "1.0",
-                        "method": "wslink.binary.attachment",
-                        "args": [key],
-                    }
-                    self.sendMessage(json.dumps(header, ensure_ascii = False).encode('utf8'))
-                    # Send binary message
-                    self.sendMessage(attachments[key], True)
-
-        self.sendMessage(encMsg)
-
-    def sendWrappedError(self, rpcid, code, message, data = None):
-        wrapper = {
-            "wslink": "1.0",
-            "id": rpcid,
-            "error": {
-                "code": code,
-                "message": message,
-            },
-        }
-        if (data):
-            wrapper["error"]["data"] = data
-        encMsg = json.dumps(wrapper, ensure_ascii = False).encode('utf8')
-        self.sendMessage(encMsg)
-
-
-    def publish(self, topic, data):
-        publishManager.publish(topic, data)
-
-    def addAttachment(self, payload):
-        return publishManager.addAttachment(payload)
-
-    def setSecret(self, newSecret):
-        self.secret = newSecret
+        print('schedule exit for {0} seconds from now')
+        schedule_callback(secondsLater, self.stopServer)
```

Comment:
Can this be canceled if someone connect?

Reply:
The diff is weird there.  That `schedule_callback` is actually the `application.exit.later` rpc method of `ServerProtocol`, where I don't think it make sense to cancel it if someone connects.